linera-sdk/src/lib.rs

```diff
@@ -265,71 +279,84 @@ pub struct QueryContext {
     /// signer.
-    pub effects: Vec<(Destination, bool, Vec<u8>)>,
+    pub effects: Vec<(Destination, bool, Effect)>,
     /// Subscribe chains to channels.
     pub subscribe: Vec<(ChannelName, ChainId)>,
     /// Unsubscribe chains to channels.
     pub unsubscribe: Vec<(ChannelName, ChainId)>,
 }
 
-impl ExecutionResult {
+impl<Effect> Default for ExecutionResult<Effect> {
+    fn default() -> Self {
+        Self {
+            effects: vec![],
+            subscribe: vec![],
+            unsubscribe: vec![],
+        }
+    }
+}
+
+impl<Effect: Serialize + std::fmt::Debug + DeserializeOwned> ExecutionResult<Effect> {
     /// Adds an effect to the execution result.
-    pub fn with_effect(
-        mut self,
-        destination: impl Into<Destination>,
-        effect: &impl Serialize,
-    ) -> Self {
-        let effect_bytes = bcs::to_bytes(effect).expect("Effect should be serializable");
-        self.effects.push((destination.into(), false, effect_bytes));
+    pub fn with_effect(mut self, destination: impl Into<Destination>, effect: Effect) -> Self {
+        self.effects.push((destination.into(), false, effect));
         self
     }
 
     /// Adds an authenticated effect to the execution result.
     pub fn with_authenticated_effect(
         mut self,
         destination: impl Into<Destination>,
-        effect: &impl Serialize,
+        effect: Effect,
     ) -> Self {
-        let effect_bytes = bcs::to_bytes(effect).expect("Effect should be serializable");
-        self.effects.push((destination.into(), true, effect_bytes));
+        self.effects.push((destination.into(), true, effect));
         self
     }
 }
 
 /// The result of calling into a user application.
-#[derive(Debug, Default, Deserialize, Serialize)]
+#[derive(Debug, Deserialize, Serialize)]
 #[cfg_attr(any(test, feature = "test"), derive(Eq, PartialEq))]
-pub struct ApplicationCallResult {
+pub struct ApplicationCallResult<Effect, Value, SessionType> {
     /// The return value.
-    #[debug(with = "linera_base::hex_debug")]
-    pub value: Vec<u8>,
+    // #[debug(with = "linera_base::hex_debug")]
+    pub value: Option<Value>,
     /// The externally-visible result.
-    pub execution_result: ExecutionResult,
+    pub execution_result: ExecutionResult<Effect>,
     /// The new sessions that were just created by the callee for us.
-    pub create_sessions: Vec<Session>,
+    pub create_sessions: Vec<Session<SessionType>>,
+}
+
+impl<Effect, Value, Session> Default for ApplicationCallResult<Effect, Value, Session> {
+    fn default() -> Self {
+        Self {
+            value: None,
+            execution_result: Default::default(),
+            create_sessions: vec![],
+        }
+    }
 }
 
 /// Syscall to request creating a new session.
 #[derive(Debug, Default, Deserialize, Serialize)]
 #[cfg_attr(any(test, feature = "test"), derive(Eq, PartialEq))]
-pub struct Session {
+pub struct Session<T> {
```

Comment:
Following your observation offline, this could be just `T = Self::SessionState` I guess.

Reply:
I'm not sure I understand this point? Are you referring to removing `kind` and therefore this entire struct becomes redundant?